gr-zeromq/lib/base_impl.cc

```diff
@@ -123,36 +123,50 @@ namespace gr {
     bool
     base_source_impl::has_pending()
     {
-      return d_msg.size() > d_consumed_bytes;
+      return !d_msgs.empty();
     }
 
     int
     base_source_impl::flush_pending(void *out_buf, const int out_nitems, const uint64_t out_offset)
     {
-      /* How much to copy in this call */
-      int to_copy_items = std::min(out_nitems, (int)((d_msg.size() - d_consumed_bytes) / d_vsize));
-      int to_copy_bytes = d_vsize * to_copy_items;
+      int total_copied_items = 0;
+      while ((total_copied_items < out_nitems) && !d_msgs.empty()) {
+        struct rx_message* msg = d_msgs.front().get();
 
-      /* Copy actual data */
-      memcpy(out_buf, (uint8_t*)d_msg.data() + d_consumed_bytes, to_copy_bytes);
+        /* How much to copy in this iteration */
+        int to_copy_items = std::min(out_nitems - total_copied_items,
+                                     (int)((msg->zmq_msg.size() - msg->consumed_bytes) / d_vsize));
+        int to_copy_bytes = d_vsize * to_copy_items;
```

Comment:
The original code had a bad assumption that to_copy_items would never be 0, as the sender and receiver were assumed to always use the same vector length, or that sender's vector length was a multiple of the receiver's vector length.  That assumption will get you stuck in an infinite loop here, I think, if the sender and receiver ever differ on vector length where that assumption is not true.

This needs to be addressed somehow.  At a minimum, prevent the infinite loop from happening.  I'd also suggest emitting some warning that sender and receiver are using incompatible vector lengths and that samples may be lost.  I'll defer to jcorgan on the specifics of notifying the user.